byway/src/main/webapp/routepage.js

```diff
@@ -12,117 +12,198 @@
- * based on what's in the schedule panel
- * @param {DirectionsService} directionsService object that communicates with the GMaps API service
- * @param {DirectionsRenderer} directionsRenderer object that renders display results on the map
- * @param {String} start starting point of route
- * @param {String} end ending point of route. TODO: In the final implementation, just have 
- * start as start and end are the same
- */
-function calcRoute(directionsService, directionsRenderer, start, end) {
+/** Displays route containing waypoints overtop the map. */
+function calcRoute() {
   let request = {
     origin:  start,
     destination: end,
-    travelMode: 'DRIVING'
+    travelMode: 'DRIVING',
+    waypoints:  route.map(waypoint => ({location: waypoint})),
+    optimizeWaypoints: true
   };
   directionsService.route(request, function(response, status) {
     if (status == 'OK') {
       directionsRenderer.setDirections(response);
+      orderWaypoints(response);
+      updateDistanceTime(response);
     } else {
       window.alert("Could not calculate route due to: " + status);
     }
+    updateRoute();
   });
 }
 
-/** Clear the stops panel in the html */
-function clearStops(){
-  const stopList = document.getElementById('stop-list');
-  if(stopList != null){
-    stopList.innerText = "";
+/** Add the starting location back to the schedule panel
+ *  TODO: Delete markers for recommended stops not selected.
+ *  TODO: Disable usage after? Don't want to keep adding to list. 
+ */
+function generateRoute() {
+  route.splice(0,0,start);
+  route.splice(route.length, 0, end);
+  console.log(route);
+  renderRouteList();
+}
+
+/**
+ * Reorders the elements in route list based on the optimized order of 
+ * waypoints returned in the response 
+ * @param {response} response response from the directions service object
+ */
+function orderWaypoints(response){
+  let waypoint_order = response.routes[0].waypoint_order;
+  let route_copy = [...route];
+  for(let i = 0; i < route.length; i++){
+    route[i] = route_copy[waypoint_order[i]];
   }
 }
 
-/** Get the new list of stops from datastore onload */
-function getStopsOnload(){
-  clearStops();
+/**
+ * Calculates and sums up the distance and time duration between all destinations (legs)
+ * @param {response} response response from the directions service object
+ * @return {number, number, number} distance total driving distance for whole route, 
+ * hours estimated driving time in hours, minutes estimated driving time in minutes
+ */
+function computeDistanceTime(response) {
+  let totalDist = 0;
+  let totalTime = 0;
+  // full route
+  let route_response = response.routes[0];
+
+  for (let i = 0; i < route_response.legs.length; i++) {
+    // in meters
+    totalDist += route_response.legs[i].distance.value;
+    // in seconds
+    totalTime += route_response.legs[i].duration.value;
+  }
+  
+  let distance = (totalDist / 1000).toFixed(2);
```

Comment:
Nit: add comment about units for distance here and in function comment.

Reply:
I have the units written in the for loop above this but I'll move it for readability, though I did list them in the function comments in the `@return`.